README.md
# Spock-Websockets

This is a simple proof-of-concept web app that demonstrates the use of [websockets](https://en.wikipedia.org/wiki/WebSocket) with the [Haskell](https://haskell.org/) [Spock](http://spock.li) library.

### Main libraries used

- [Spock](http://spock.li), a lightweight [Haskell](https://haskell.org/) web framework
- [websockets](https://hackage.haskell.org/package/websockets), a sensible and clean way to write WebSocket-capable servers in Haskell

### Demo

The source code contains a simple demo that streams the sequence of numbers that are displayed in the browser with a visual gimmick.

![animated gif](spock-websockets.gif)

### How it works

The key thing to realize is that [websocketsOr](https://hackage.haskell.org/package/wai-websockets-3.0.1.2/docs/Network-Wai-Handler-WebSockets.html#v:websocketsOr) produces a [Middleware](https://hackage.haskell.org/package/wai-3.2.1.2/docs/Network-Wai.html#t:Middleware) that can be supplied to [runSpock](https://spockdocs.s3.eu-central-1.amazonaws.com/Spock-0.12.0.1/Web-Spock.htm#v:runSpockl).

#### Middleware

Middleware is a component that sits between the server and application:

```haskell
type Middleware = Application -> Application	-- defined in Network.Wai
```

and `websocketsOr` can produce one

```haskell
wsMiddleware :: Middleware
wsMiddleware =  websocketsOr defaultConnectionOptions wsApp
    where ...
```

The `ServerApp` is 

```haskell
wsApp :: ServerApp
	wsApp pendingConn = do
	conn <- acceptRequest pendingConn
	counter conn 1
```

#### Counter

The `counter` is an endless recursion (loop) that just counts up:

```haskell
counter :: Connection -> Int -> IO ()
counter conn i = do
    threadDelay 50000   -- 50ms
    sendTextData conn (T.pack $ show i)
    counter conn (i + 1)
```

This is tied in with the Spock app by defining the `appMiddlewares`

```haskell
appMiddlewares = do
    middleware (staticPolicy $ noDots >-> addBase "static")
    middleware wsMiddleware
```

#### Start Spock

Now all is ready to start the server:

```haskell
runSpock 8080 (spock spockConfig (appMiddlewares >> app))
```



ChangeLog.md
# Changelog for test-sp

## Unreleased changes


static/ws.js
window.onload = function() {
    console.log("window.onload");
    var ws = createWebSocket('/');

    ws.onopen = function() {
        console.log('onopen');
    };

    ws.onmessage = function (event) {
        var i = Number(event.data);
        var counter = document.getElementById("counter");
        counter.innerHTML = i;
        var boxes10 = document.getElementById("boxes10");
        if(i % 10 == 0) {
            boxes10.innerHTML = '';
            var boxes1 = document.getElementById("boxes1");
            if(i % 100 == 0) {
                boxes1.innerHTML = '';
            } else {
                var div = document.createElement("div"); 
                div.className = "box";
                boxes1.appendChild(div);
            }
        } else {
            var div = document.createElement("div"); 
            div.className = "box";
            boxes10.appendChild(div);
        }
    }
}

function createWebSocket(path) {
    var host = window.location.hostname;
    if(host == '') host = 'localhost';
    var uri = 'ws://' + host + ':8080' + path;
    return new WebSocket(uri);
}

Main.hs
{-# LANGUAGE DataKinds         #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE TypeOperators     #-}
module Main where

import           Control.Concurrent                   (forkIO, threadDelay)
import           Control.Exception                    (AsyncException,
                                                       fromException, handle,
                                                       throwIO)
import           Control.Monad                        (void)
import           Data.Text                            (Text)
import qualified Data.Text                            as T
import qualified Data.Text.Lazy                       as TL
import           Network.Wai                          (Application, Middleware)
import           Network.Wai.Handler.WebSockets
import           Network.Wai.Middleware.RequestLogger (logStdoutDev)
import           Network.Wai.Middleware.Static
import           Network.WebSockets
import           Text.Blaze.Html.Renderer.Text        (renderHtml)
import           Text.Blaze.Html5                     ((!))
import qualified Text.Blaze.Html5                     as H
import qualified Text.Blaze.Html5.Attributes          as A
import           Web.Spock                            as Sp
import           Web.Spock.Config                     as Sp
import           Web.Spock.Core                       (SpockCtxT)

data MySession = EmptySession
type MyAppState = ()

main :: IO ()
main = do
    spockConfig <- defaultSpockCfg EmptySession PCNoDatabase ()
    runSpock 8080 (spock spockConfig (appMiddlewares >> app))

appMiddlewares :: Web.Spock.Core.SpockCtxT () (WebStateM () MySession MyAppState) ()
appMiddlewares = do
    middleware logStdoutDev                                     -- add logging
    middleware (staticPolicy $ noDots >-> addBase "static")     -- add static files
    middleware wsMiddleware                                     -- add websocket

wsMiddleware :: Middleware
wsMiddleware =  websocketsOr defaultConnectionOptions wsApp
    where
        wsApp :: ServerApp
        wsApp pendingConn = do
            conn <- acceptRequest pendingConn
            forkPingThread conn 30  -- not needed here, but good to have in most other situations
            sendTextData conn ("..." :: Text)
            counter conn 1
        counter :: Connection -> Int -> IO ()
        counter conn i = do
            threadDelay 50000   -- 50ms
            sendTextData conn (T.pack $ show i)
            counter conn (i + 1)

app :: SpockM () MySession MyAppState ()
app = get root action   -- there's only one route in this app

content :: H.Html
content =
    H.docTypeHtml ! A.lang "en" $ do
        H.head $ do
            H.meta ! A.charset "utf-8"
            H.title "Spock Websockets"
            H.script ! A.type_ "text/javascript" ! A.src "ws.js" $ mempty
            H.link ! A.type_ "text/css" ! A.rel "stylesheet" ! A.href "ws.css"
        H.body $ do
            H.h1 "Spock Websockets"
            H.p $ do
                "Counter: "
                H.span ! A.id "counter" $ mempty
            H.p ! A.id "boxes1" ! A.class_ "boxes" $ mempty
            H.p ! A.id "boxes10" ! A.class_ "boxes" $ mempty

action :: ActionCtxT ctx (WebStateM () MySession MyAppState) a
action = (Sp.html . TL.toStrict . renderHtml) content